server/src/main/java/de/uftos/repositories/solver/SolverRepositoryImpl.java

```diff
@@ -1,25 +1,518 @@
+      TimetableProblemDto timetable) {
+    HashMap<String, GradeTimefoldInstance> grades = new HashMap<>();
+    HashMap<String, RoomTimefoldInstance> rooms = new HashMap<>();
+    HashMap<String, StudentGroupTimefoldInstance> studentGroups = new HashMap<>();
+    HashMap<String, StudentTimefoldInstance> students = new HashMap<>();
+    HashMap<String, SubjectTimefoldInstance> subjects = new HashMap<>();
+    HashMap<String, TeacherTimefoldInstance> teachers = new HashMap<>();
+    HashMap<String, TimeslotTimefoldInstance> timeslots = new HashMap<>();
+    HashMap<String, TagTimefoldInstance> tags = new HashMap<>();
+    HashMap<String, LessonTimefoldInstance> lessons = new HashMap<>();
+    List<ConstraintInstanceTimefoldInstance> constraintInstances = new ArrayList<>();
+
+    TimetableSolutionTimefoldInstance solution = new TimetableSolutionTimefoldInstance();
+    HashMap<String, ResourceTimefoldInstance> resources = new HashMap<>();
+    resources.put("this", solution);
+
+    HashMap<String, ConstraintDefinitionTimefoldInstance> definitions = new HashMap<>();
+
+    for (ConstraintDefinitionDto definition : timetable.definitions()) {
+      definitions.put(definition.name(),
+          ConstraintDefinitionFactory.getConstraintDefinition(definition));
+    }
+
+    //initializing all resources
+    for (ResourceProblemDto resource : timetable.getResources()) {
+      switch (resource.getType()) {
+        case GRADE -> grades.put(resource.getId(), new GradeTimefoldInstance(resource.getId()));
+        case LESSON -> lessons.put(resource.getId(), new LessonTimefoldInstance(resource.getId()));
+        case ROOM -> rooms.put(resource.getId(), new RoomTimefoldInstance(resource.getId()));
+        case STUDENT_GROUP ->
+            studentGroups.put(resource.getId(), new StudentGroupTimefoldInstance(resource.getId()));
+        case STUDENT ->
+            students.put(resource.getId(), new StudentTimefoldInstance(resource.getId()));
+        case SUBJECT ->
+            subjects.put(resource.getId(), new SubjectTimefoldInstance(resource.getId()));
+        case TAG -> tags.put(resource.getId(), new TagTimefoldInstance(resource.getId()));
+        case TEACHER ->
+            teachers.put(resource.getId(), new TeacherTimefoldInstance(resource.getId()));
+        case TIMESLOT ->
+            timeslots.put(resource.getId(), new TimeslotTimefoldInstance(resource.getId()));
+        default -> throw new IllegalStateException();
+      }
+    }
+
+    //connecting all resources as defined by the Dtos (and checking connection)
+
+    for (GradeProblemDto grade : timetable.grades()) {
+      GradeTimefoldInstance timefoldInstance = grades.get(grade.id());
+      for (String studentGroupId : grade.studentGroupIds()) {
+        StudentGroupTimefoldInstance studentGroup = studentGroups.get(studentGroupId);
+        if (studentGroup.getGrade() != null) {
+          throw new IllegalArgumentException();
+        }
```

Comment:
In case of data-inconsistencies the grade might reference a student group which doesn't exist.
In this case the HashMap would return null.
If the data is consistent this IllegalArgumentException should not be thrown.

Reply:
So there could be a potential inconsistency against which you tried to harden the code, makes sense!